File: app/controllers/Streams.java
package controllers;

import models.TContact;
import models.TPost;
import models.User;
import play.mvc.Controller;
import play.mvc.Result;
import service.post.Redis;
import service.post.Receiver;
import service.post.StreamComet;

import java.util.ArrayList;
import java.util.List;

public class Streams extends Controller {

    public static Result list(){
        final User user = Application.getLocalUser(session());
        int pageNum = request().getQueryString("pageNum")==null?1:Integer.parseInt(request().getQueryString("pageNum"));
        int maxRow = request().getQueryString("maxRow")==null?1:Integer.parseInt(request().getQueryString("maxRow"));
        List<TPost> posts = TPost.findPublics(user,pageNum,maxRow);
        return ok();
    }

    public static Result news() {
        final User user = Application.getLocalUser(session());
        if(user == null)
            return ok();
        String cometJS = request().getQueryString("callback");
        List<String> channels = new ArrayList<String>();
        for(TContact contact : user.self_contacts){
            channels.add(Redis.personChannel(contact.person.id));
        }
        StreamComet client = new StreamComet(cometJS,user,channels);
        Thread receiveThread = new Thread(new Runnable() {
            @Override
            public void run() {
            Receiver receiver = new Receiver();
            for(TContact contact : user.self_contacts){
                Redis.subscribe(receiver, Redis.personChannel(contact.person.id));
            }
            }
        });
        receiveThread.start();
        return ok(client);
    }

}

File: app/controllers/Streams.scala
package controllers;

import play.api.mvc._
import models._
import views._
import com.mongodb.casbah.commons.MongoDBObject
import service._
import play.api.libs.iteratee.Enumerator
import java.util.concurrent.TimeUnit
import akka.pattern.ask
import akka.util.Timeout
import play.api.libs.concurrent.Execution.Implicits._
import play.api.libs.Comet
import securesocial.core.SecureSocial
import scala.concurrent._


object Streams extends Controller with SecureSocial {

  def comet = UserAwareAction { request =>
    AsyncResult {
      request.user match {
        case Some(user) => {
          val cometJS = request.getQueryString("callback").getOrElse("")
          for(followed <- User.findContactUser(user)){
            Redis.subscribe("user_"+user.identityId)
          }
          implicit val timeout = Timeout(5,TimeUnit.SECONDS)
          val cometFuture = (CometActor.ref ? Connect(user,cometJS)).mapTo[Enumerator[String]]
          cometFuture.map { chunks =>
            Console.println("***********"+chunks)
            Ok.stream(chunks >>> Enumerator.eof).as(JAVASCRIPT)
          }
        }
		case None => Future.successful(NotFound)
      }

    }
  }

  def markread = Action { request =>
    Ok
  }

}
